### PR TITLE
Optimize ecma_op_function_call

### PR DESCRIPTION
**Pull request by rerobika**
Remove redundant isCallable checks

JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik robert.fancsik@h-lab.eu

### CODE REVIEW

File: jerry-core/ecma/operations/ecma-function-object.c [hunk 1]
```diff
@@ -1390,27 +1454,47 @@ ecma_op_function_call (ecma_object_t *func_obj_p, /**< Function object */
...
-  if (JERRY_LIKELY (type == ECMA_OBJECT_TYPE_FUNCTION))
-  {
-    result = ecma_op_function_call_simple (func_obj_p, this_arg_value, arguments_list_p, arguments_list_len);
-  }
-  else if (type == ECMA_OBJECT_TYPE_BUILT_IN_FUNCTION)
-  {
-    result = ecma_op_function_call_native_built_in (func_obj_p, this_arg_value, arguments_list_p, arguments_list_len);
-  }
-  else if (type == ECMA_OBJECT_TYPE_NATIVE_FUNCTION)
+  switch (ecma_get_object_type (func_obj_p))
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-function-object.c:1457]: The original code was in likelyhood order. Is this faster?

File: jerry-core/ecma/operations/ecma-function-object.h [hunk 2]
```diff
@@ -91,6 +94,10 @@ ecma_op_get_prototype_from_constructor (ecma_object_t *ctor_obj_p, ecma_builtin_
 ecma_value_t
 ecma_op_function_has_instance (ecma_object_t *func_obj_p, ecma_value_t value);
 
+ecma_value_t
+ecma_op_function_callv (ecma_value_t callee,  ecma_value_t this_arg_value,  const ecma_value_t *arguments_list_p,
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-function-object.h:98]: ecma_op_function_validated_call

**Review by zherczeg — CHANGES_REQUESTED** [review]

**Comment by rerobika** [comment]
Benchmark | Heap (bytes) | Stack (bytes) | Perf (sec) |
----: | ----: | ----: | ----: |
0b1100oclock.js | 7648 -> 7648 : 0.000% | 1568 -> 1568 : 0.000% | 0.287 -> 0.286 : +0.220% |
3d-cube.js | 10992 -> 10992 : 0.000% | 14280 -> 14280 : 0.000% | 1.101 -> 1.102 : -0.088% |
3d-morph.js | 289248 -> 289248 : 0.000% | 1576 -> 1576 : 0.000% | 1.328 -> 1.316 : +0.856% |
3d-raytrace.js | 119456 -> 119456 : 0.000% | 2572 -> 2572 : 0.000% | 1.286 -> 1.273 : +1.052% |
FitbitOS-Really-Basic.js | 2048 -> 2048 : 0.000% | 1504 -> 1504 : 0.000% | 0.559 -> 0.550 : +1.667% |
Flashlight-App-for-FitBit-Ionic.js | 7160 -> 7160 : 0.000% | 1992 -> 1992 : 0.000% | 0.994 -> 0.973 : +2.071% |
access-binary-trees.js | 27120 -> 27120 : 0.000% | 3548 -> 3548 : 0.000% | 0.750 -> 0.751 : -0.086% |
access-fannkuch.js | 1816 -> 1816 : 0.000% | 1648 -> 1648 : 0.000% | 2.519 -> 2.542 : -0.921% |
access-nbody.js | 4592 -> 4592 : 0.000% | 1736 -> 1736 : 0.000% | 1.442 -> 1.451 : -0.610% |
bitops-3bit-bits-in-byte.js | 1048 -> 1048 : 0.000% | 1592 -> 1592 : 0.000% | 0.701 -> 0.686 : +2.139% |
bitops-bits-in-byte.js | 1008 -> 1008 : 0.000% | 1592 -> 1592 : 0.000% | 0.998 -> 0.987 : +1.061% |
bitops-bitwise-and.js | 784 -> 784 : 0.000% | 1344 -> 1344 : 0.000% | 1.248 -> 1.251 : -0.244% |
bitops-nsieve-bits.js | 61208 -> 61208 : 0.000% | 1592 -> 1592 : 0.000% | 1.477 -> 1.486 : -0.646% |
controlflow-recursive.js | 1288 -> 1288 : 0.000% | 61584 -> 61584 : 0.000% | 0.510 -> 0.508 : +0.276% |
crypto-aes.js | 28144 -> 28144 : 0.000% | 2308 -> 2308 : 0.000% | 0.884 -> 0.878 : +0.703% |
crypto-md5.js | 70896 -> 70896 : 0.000% | 1720 -> 1720 : 0.000% | 0.724 -> 0.720 : +0.595% |
crypto-sha1.js | 46240 -> 46240 : 0.000% | 1720 -> 1720 : 0.000% | 0.745 -> 0.737 : +1.025% |
crypto.js | 51192 -> 51192 : 0.000% | 3196 -> 3196 : 0.000% | 8.549 -> 8.567 : -0.215% |
date-format-tofte.js | 8840 -> 8840 : 0.000% | 2016 -> 2016 : 0.000% | 0.986 -> 0.983 : +0.253% |
date-format-xparb.js | 15200 -> 15200 : 0.000% | 2576 -> 2576 : 0.000% | 0.731 -> 0.716 : +2.020% |
deltablue.js | 456312 -> 456312 : 0.000% | 3592 -> 3592 : 0.000% | 5.334 -> 5.245 : +1.668% |
fibit-yoga-app.js | 6136 -> 6136 : 0.000% | 2544 -> 2544 : 0.000% | 2.090 -> 2.061 : +1.428% |
fitbit-week-number.js | 1544 -> 1544 : 0.000% | 1656 -> 1656 : 0.000% | 0.357 -> 0.354 : +0.777% |
ionic-watchface-borderlands.js | 8720 -> 8720 : 0.000% | 1692 -> 1692 : 0.000% | 2.025 -> 1.991 : +1.675% |
jacadi.js | 4088 -> 4088 : 0.000% | 7868 -> 7868 : 0.000% | 1.600 -> 1.553 : +2.951% |
math-cordic.js | 2064 -> 2064 : 0.000% | 1592 -> 1592 : 0.000% | 1.652 -> 1.659 : -0.381% |
math-partial-sums.js | 1528 -> 1528 : 0.000% | 1568 -> 1568 : 0.000% | 1.136 -> 1.141 : -0.472% |
math-spectral-norm.js | 3248 -> 3248 : 0.000% | 1592 -> 1592 : 0.000% | 0.750 -> 0.753 : -0.357% |
raytrace.js | 22008 -> 22008 : 0.000% | 5168 -> 5168 : 0.000% | 2.618 -> 2.622 : -0.131% |
richards.js | 8472 -> 8472 : 0.000% | 2156 -> 2156 : 0.000% | 0.300 -> 0.295 : +1.812% |
sdk-altimeter.js | 1632 -> 1632 : 0.000% | 1736 -> 1736 : 0.000% | 2.306 -> 2.190 : +5.008% |
sdk-hr-meter.js | 1688 -> 1688 : 0.000% | 1692 -> 1692 : 0.000% | 0.230 -> 0.220 : +4.174% |
sdk-lcd-clock.js | 5624 -> 5624 : 0.000% | 2048 -> 2048 : 0.000% | 1.575 -> 1.534 : +2.613% |
string-base64.js | 90120 -> 90120 : 0.000% | 1640 -> 1640 : 0.000% | 1.442 -> 1.404 : +2.625% |
string-fasta.js | 4400 -> 4400 : 0.000% | 1576 -> 1576 : 0.000% | 2.798 -> 2.790 : +0.285% |
Geometric mean: | 8685.825 -> 8685.825 : 0.000% | 2377 -> 2377 : 0.000% | 1.132 -> 1.12 : +1.060% |

Binary (bytes) | master(7ea0000ee0) | patch(84d9a82728) | Diff |
----: | ----: | ----: | ----: |
size | 235132 | 235132 | 0 bytes |
.rodata | 20878 | 20878 | 0 bytes |
.dynstr | 384 | 384 | 0 bytes |
.rel.plt | 384 | 384 | 0 bytes |
.interp | 25 | 25 | 0 bytes |
.dynsym | 864 | 864 | 0 bytes |
.gnu.hash | 428 | 428 | 0 bytes |
.text | 204760 | 204752 | -8 bytes |
.comment | 85 | 85 | 0 bytes |
.shstrtab | 226 | 226 | 0 bytes |
.data | 8 | 8 | 0 bytes |
.ARM.exidx | 8 | 8 | 0 bytes |
.rel.dyn | 16 | 16 | 0 bytes |
.init | 12 | 12 | 0 bytes |
.got | 208 | 208 | 0 bytes |
.plt | 608 | 608 | 0 bytes |
.note.ABI-tag | 32 | 32 | 0 bytes |
.gnu.version_r | 64 | 64 | 0 bytes |
.bss | 1575136 | 1575136 | 0 bytes |
.fini | 8 | 8 | 0 bytes |
.hash | 372 | 372 | 0 bytes |
.gnu.version | 108 | 108 | 0 bytes |
.fini_array | 4 | 4 | 0 bytes |
.init_array | 4 | 4 | 0 bytes |
.dynamic | 248 | 248 | 0 bytes |
.eh_frame | 4 | 4 | 0 bytes |
.ARM.attributes | 53 | 53 | 0 bytes |

File: jerry-core/ecma/operations/ecma-function-object.c (hunk 1 above)
Comment (rerobika) [reply jerry-core/ecma/operations/ecma-function-object.c:1457]: Based on the benchmark results below it is.

File: jerry-core/ecma/operations/ecma-function-object.h (hunk 2 above)
Comment (rerobika) [reply jerry-core/ecma/operations/ecma-function-object.h]: Updated.

File: jerry-core/ecma/operations/ecma-function-object.c [hunk 3]
```diff
@@ -1022,6 +1039,76 @@ ecma_op_get_prototype_from_constructor (ecma_object_t *ctor_obj_p, /**< construc
...
+#if JERRY_ESNEXT
+/**
+ * Perform a JavaScript class function object method call.
+ *
+ * The input function object should be a JavaScript class constructor
+ *
+ * @return the result of the function call.
+ */
+static ecma_value_t JERRY_ATTR_NOINLINE
+ecma_op_function_call_ctor (vm_frame_ctx_shared_args_t *shared_args_p, /**< shared data */
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-function-object.c:1051]: ecma_op_function_call_constructor

File: jerry-core/vm/vm-stack.c [hunk 4]
```diff
@@ -380,60 +380,53 @@ vm_stack_find_finally (vm_frame_ctx_t *frame_ctx_p, /**< frame context */
...
-              if (!is_object)
-              {
-                result = ecma_raise_type_error (ECMA_ERR_MSG ("Iterator 'return' result is not object"));
-              }
+            if (!is_object)
+            {
+              result = ecma_raise_type_error (ECMA_ERR_MSG ("Iterator 'return' result is not object"));
             }
           }
+
```
Comment (zherczeg) [inline jerry-core/vm/vm-stack.c:429]: Remove this newline

**Review by zherczeg — APPROVED** [review]
LGTM

**Review by dbatyai — APPROVED** [review]
LGTM

**Pull request merged**